### ISSUE TITLE
Version 2.341+ (up to current 2.405) breaks use case to trigger two shortcuts at once

### ISSUE THREAD

**Issue by craftey** [fifafu/BetterTouchTool#2270]
Newer versions 2.341+ (currently up to 2.405) have introduced a change how keyboard shortcuts work. Unfortunately this breaks one of my global keyboard shortcut configurations.

I have a two monitor setup. On each monitor I have 2 Missioncontrol-Desktops. So Monitor 1 gets desktop 1 and 2. Monitor 2 gets desktop 3 and 4. I want to change both desktop on both monitors with one short cut. So I configured the following:

![bildschirmfoto 2018-03-16 um 12 18 24](https://user-images.githubusercontent.com/17045396/37554670-69b43968-29dc-11e8-93d0-d30ca09e3613.png)

This works perfectly with version 2.340. But versions after that seem not to send both short cuts  anymore e.g. when hitting OPT+1 it used to trigger OPT+1 and OPT+3. I also tried to send OPT+3 after a delay. This is hidden in the above picture under the unexpanded arrow on the left.

No matter what I set for `"Use old keyboard short implementation"` in newer versions, I cannot get the old behavior to work again.

Kindly,
Craftey

#### Device information:

Type of Mac: MacBook Pro 13 Touchbar + external magic keyboard
macOS version: 10.13.3
BetterTouchTool version: 2.405

**Comment by craftey** [comment]
I found a workaround tested with current version 2.405:
<img width="677" alt="bildschirmfoto 2018-03-17 um 12 55 00" src="https://user-images.githubusercontent.com/17045396/37555118-b887860c-29e2-11e8-8803-9c5cc4cd457c.png">
**Note:** You must enable "Prevent recursive triggers" for this short cut config.

Interestingly the last column with the enable/disable flag shows that I can disable the delay, it does not change the behavior of the little program. What effectively breaks the program is setting the delay to 0.15 or smaller. Just disabling the delay has no effect it seems alway on. But setting the delay to a small enough value demonstrates, that the delay is needed.

**Comment by fifafu** [comment]
I have made some changes in recent alphas, also I have added "Switch to Desktop X" predefined actions (and "move window to desktop x").

Maybe you could check whether that changes the behavior you observed (v2.409, get via "check for alpha version updates")

**Comment by craftey** [comment]
Btw, I am not quite sure about my analysis. On a different machine, an older one, with Mac OS El Captain the said workaround does not do the trick.

I would like to have at look at the newer versions. Unfortunately the sparkle update for BTT is broken on my machine. The reason ist not clear so far. So the only way to test things atm would be via direct downloads.

![bildschirmfoto 2018-03-13 um 08 19 01](https://user-images.githubusercontent.com/17045396/37327652-80f2fa6c-2697-11e8-9259-dffa115d8030.png)
Would be great, if we could debug this update issue somehow.

**Comment by craftey** [comment]
Here the log from Console.app when trying to update via sparkle:
`standard	20:13:19.655192 +0100	BetterTouchTool	Extracting using '/usr/bin/ditto' '-x' '-k' '-' < '/Users/<username>/Library/Caches/com.hegenberg.BetterTouchTool/org.sparkle-project.Sparkle/BetterTouchTool 760/btt2.410.zip' '/Users/<username>/Library/Caches/com.hegenberg.BetterTouchTool/org.sparkle-project.Sparkle/BetterTouchTool 760'
standard	20:13:21.616829 +0100	BetterTouchTool	Task <0E551303-399E-4083-9AB8-6A88444442AD>.<42> now using Connection 27
standard	20:13:21.619526 +0100	BetterTouchTool	Task <0E551303-399E-4083-9AB8-6A88444442AD>.<42> sent request, body S
standard	20:13:21.796039 +0100	BetterTouchTool	Task <0E551303-399E-4083-9AB8-6A88444442AD>.<42> received response, status 200 content K
standard	20:13:21.796248 +0100	BetterTouchTool	Task <0E551303-399E-4083-9AB8-6A88444442AD>.<42> response ended
standard	20:13:22.304209 +0100	BetterTouchTool	Task <6E859A5D-42A1-43C6-9B3E-D486E2423542>.<43> now using Connection 27
standard	20:13:22.309718 +0100	BetterTouchTool	Task <6E859A5D-42A1-43C6-9B3E-D486E2423542>.<43> sent request, body S
standard	20:13:22.516611 +0100	BetterTouchTool	Task <6E859A5D-42A1-43C6-9B3E-D486E2423542>.<43> received response, status 200 content K
standard	20:13:22.516877 +0100	BetterTouchTool	Task <6E859A5D-42A1-43C6-9B3E-D486E2423542>.<43> response ended
standard	20:13:24.806769 +0100	BetterTouchTool	Task <1E1DDC18-1897-4606-9821-9798A8DBB56D>.<44> now using Connection 27
standard	20:13:24.810258 +0100	BetterTouchTool	Task <1E1DDC18-1897-4606-9821-9798A8DBB56D>.<44> sent request, body S
standard	20:13:24.970174 +0100	BetterTouchTool	Task <1E1DDC18-1897-4606-9821-9798A8DBB56D>.<44> received response, status 200 content K
standard	20:13:24.970411 +0100	BetterTouchTool	Task <1E1DDC18-1897-4606-9821-9798A8DBB56D>.<44> response ended
standard	20:13:25.963546 +0100	BetterTouchTool	Task <735F0FEE-BED4-4406-BD31-B36781D672FF>.<45> now using Connection 27
standard	20:13:25.972166 +0100	BetterTouchTool	Task <735F0FEE-BED4-4406-BD31-B36781D672FF>.<45> sent request, body S
standard	20:13:26.116803 +0100	BetterTouchTool	Task <735F0FEE-BED4-4406-BD31-B36781D672FF>.<45> received response, status 200 content K
standard	20:13:26.116946 +0100	BetterTouchTool	Task <735F0FEE-BED4-4406-BD31-B36781D672FF>.<45> response ended
standard	20:13:26.693298 +0100	BetterTouchTool	CSSM Exception: -2147415792 CSSMERR_CSP_INVALID_KEY
`
The last line seems relevant, because it appeared as soon as I hit the restart now button.

Thanks in advance for any help.

**Comment by craftey** [comment]
Trashing the folder `~/Library/Caches/com.hegenberg.BetterTouchTool` fixed the sparkle updater. 😊

**Comment by fifafu** [comment]
Very interesting!
I assume somehow the permissions of that folder got corrupted and caused sparkle to be unable to download the new file to the cache folder. Then when it tried to relaunch the code signature check failed because it was still looking at the old version.

**Comment by craftey** [comment]
I have tested 2.410 on 10.13.3 and on El Capitan. The old behavior with double short cuts seems back, at least on 10.13.3. On El Capitan I could not get the double short cuts triggered reliably.

The `Switch to Desktop x` feature looks also nice and seems to work for my use case on 10.13.3. Though it is not available on El Capitan.

The option `"Use old keyboard short implementation"` seems gone in 2.410 am I right?

**Comment by fifafu** [comment]
@craftey, thanks for reporting the El Capitan issues (it's definitely possible that the new APIs are not supported there). I'll test that soon.

The old shortcut implementation is now in a separate "Keyboard" tab in the advanced settings.

**Comment by craftey** [comment]
> ... and caused sparkle to be unable to download the new file to the cache folder. ...

From my observation of the cache folder, I can verify that the download succeeded. In the above log we see: `/Users/<username>/Library/Caches/com.hegenberg.BetterTouchTool/org.sparkle-project.Sparkle/BetterTouchTool 760/btt2.410.zip`
I first wanted to manually install that file. But than I decided to trash the cache and that fixed it.
Now after the successful update I see, that the folder `/Users/<username>/Library/Caches/com.hegenberg.BetterTouchTool/org.sparkle-project.Sparkle/` is empty. That was not the case, when the update always failed. So it seems the cleanup is a mandatory step, which got skipped somehow earlier.

**Comment by craftey** [comment]
I got the double short cuts working on El Capitain on BTT 2.410. In some of my tests I had forgotten to set `"Prevent recursive triggers"` and therefore got bad results. Now this config seems to work on El Cap:

**Comment by craftey** [comment]
You are quick in closing issues. But I understand this. :)

Let me add some followup: I said I wanted to test this issue with the dual monitor setup at my workplace. I did so.
What I found is a bug in High Sierra ☹️.

First of all, switching two desktops on two different monitors with the versions 2.410+ works now on both High Sierra and El Capitan. The following setup does the trick:
![bildschirmfoto 2018-03-26 um 11 34 59](https://user-images.githubusercontent.com/17045396/37898713-d2719c7a-30e9-11e8-9c44-ff4cabd54059.png)

On High Sierra following also works:
![bildschirmfoto 2018-03-21 um 15 29 40](https://user-images.githubusercontent.com/17045396/37898739-e6865dcc-30e9-11e8-836e-ca4e2fbb2dc8.png)

But if we have Mission Control (aka Expose) open then on both MacOS versions the above setup does not switch desktops on both monitors. As I said on El Captain the following is a work around:
![bildschirmfoto](https://user-images.githubusercontent.com/17045396/37560467-d2ea6306-2a38-11e8-9296-1b2338d4a91b.png)

Though, this work around does not work on High Sierra. Also the following does not work:
![bildschirmfoto 2018-03-21 um 15 47 55](https://user-images.githubusercontent.com/17045396/37898852-33465216-30ea-11e8-9342-340ed6c14965.png)

I found that even when *BTT is shut down* and Mission Control (aka Expose) is open, then it depends on which monitor the mouse pointer is actually positioned on. If it is on monitor 1, than `OPT+1`/`OPT+2` change the desktops on monitor 1. If the mouse pointer is on on monitor 2, than `OPT+1`/`OPT+2` change the desktops on monitor 2. In both cases `OPT+3`/`OPT+4` does not change any desktop. As opposed to on El Capitan this works as expected.

It seems that I have to live with this minor downgrade in functionality on High Sierra. Except BTT can do anything to mitigate the High Sierra bug 🤓 🙏. Maybe a mouse pointer hack like with action "Mission Control & Immediately Show Desktop Preview".

Kindly, craftey.
